external/vcm/vcm/derived_mapping.py

```diff
@@ -379,3 +379,55 @@ def upward_heat_flux_at_surface(self):
     return result.assign_attrs(
         long_name="Upward heat (sensible+radiative) flux at surface", units="W/m**2"
     )
+
+
+@DerivedMapping.register(
+    "incloud_water_mixing_ratio",
+    required_inputs=["cloud_amount", "cloud_water_mixing_ratio"],
+    use_nonderived_if_exists=True,
+)
+def incloud_water_mixing_ratio(self):
+    result = vcm.gridcell_to_incloud_condensate(
+        self["cloud_amount"], self["cloud_water_mixing_ratio"]
+    )
+    return result.assign_attrs(long_name="in-cloud water mixing ratio", units="kg/kg")
+
+
+@DerivedMapping.register(
+    "incloud_ice_mixing_ratio",
+    required_inputs=["cloud_amount", "cloud_ice_mixing_ratio"],
+    use_nonderived_if_exists=True,
+)
+def incloud_ice_mixing_ratio(self):
+    result = vcm.gridcell_to_incloud_condensate(
+        self["cloud_amount"], self["cloud_ice_mixing_ratio"]
+    )
+    return result.assign_attrs(long_name="in-cloud ice mixing ratio", units="kg/kg")
+
+
+@DerivedMapping.register(
+    "gridcell_cloud_water_mixing_ratio",
+    required_inputs=["cloud_amount", "incloud_water_mixing_ratio"],
+)
+def gridcell_cloud_water_mixing_ratio(self):
+    # needs a nonstandard name to avoid circular reference
```

Comment:
Could you say more about this? Feels like this may be pointing to a design issue with `vcm.DerivedMapping`.

Reply:
The DerivedMapping doesn't allow a set of conversions where 'a' depends on 'b' and 'b' depends on 'a', since it recurses to find the dependency graph of everything in the mapping (it's a DAG). I think it's fine, and in my case I think it's useful to have a different name to differentiate the output of these conversions from a standard mixing ratio field.

Reply:
In the case of variables that could be obtained in different ways I think the `TransformedPredictor` may be a more suitable approach, e.g. there's a common naming pattern for transforms there that includes how the quantity was derived like `Q1_from_...`.

Reply:
I'm fine either way. I agree this feels like a use-case that the `TransformedPredictor` is naturally suited for. But it also requires an additional configuration burden which is a downside. If there is something meaningfully different about the differently named diagnostics, then maybe it is okay.

Reply:
@AnnaKwa and I just discussed this and came to the conclusion that the conversion to "in-cloud" is unambiguous from wrapper variables and is appropriate for the `DerivedMapper`. The conversion back to grid-cell means from in-cloud is more problem-specific and also creates the potential circular reference, so it should be moved to a `DataTransform`. This has the nice result of only needing additional config on the output variables side, not on the training data.